### ISSUE TITLE
Microservices do not accept content type "application/x-www-form-urlencoded;charset=utf-8"

### ISSUE THREAD

**Issue by kleegrewec** [frmichel/sparql-micro-service#20]
I have implemented several sparql microservices which work fine in case I access them directly using curl: e.g. 

``````
adminuser@dev-sag-jumphost:~/bin$ curl -vvvv  --header "Content-Type:application/x-www-form-urlencoded" --data-urlencode 'query=prefix sim:<http://ontology.siemens.com/si/measurement/> SELECT * WHERE { ?measurement a sim:Measurement ; sim:time_stamp "22.05.10T10:00:00.000" ; ?p ?o. } ' http://dev-sag-sms.ensure2.siemens.cloud/service/siguard-pdp-sd/getAllChannelsByTimestamp
*   Trying 10.0.7.181...
* TCP_NODELAY set
* Connected to dev-sag-sms.ensure2.siemens.cloud (10.0.7.181) port 80 (#0)
> POST /service/siguard-pdp-sd/getAllChannelsByTimestamp HTTP/1.1
> Host: dev-sag-sms.ensure2.siemens.cloud
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Type:application/x-www-form-urlencoded
> Content-Length: 241
>
* upload completely sent off: 241 out of 241 bytes
< HTTP/1.1 200 OK
< Date: Wed, 24 Aug 2022 08:21:30 GMT
< Server: Apache/2.4.52 (Debian)
< X-Powered-By: PHP/7.4.27
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
< Access-Control-Allow-Origin: *
< Content-Type: application/sparql-results+xml
< Set-Cookie: PHPSESSID=46snsnrqnfs42g6pbqhn0sn8s5; path=/
< Transfer-Encoding: chunked
<
<?xml version="1.0" ?>
<sparql xmlns='http://www.w3.org/2005/sparql-results#'>
<head>
<variable name='measurement'/>
<variable name='p'/>
<variable name='o'/>
</head>
<results>
<result>
<binding name='measurement'><uri>http://ontology.siemens.com/si/measurement/meas_1012_22.05.10T10:00:00.000</uri></binding>
<binding name='p'><uri>http://ontology.siemens.com/si/measurement/has_angle</uri></binding>
<binding name='o'><literal datatype='http://www.w3.org/2001/XMLSchema#double'>0.00929626</literal></binding>
</result>
...
``````

But when I use federated queries to a blazegraph sparql endpoint using a service clause the query fails with the error:

Response=Erroneous request: SPARQL query with HTTP POST method but **unexpected 'Content-Type': application/x-www-form-urlencoded; charset=UTF-8**

It seems that the blazegraph adds the charset information to the call to service but there seem to be no way to stop this. I can verify the behavior if I add the charset info to my direct query:

``````
adminuser@dev-sag-jumphost:~/bin$ curl -vvvv  --header "Content-Type:application/x-www-form-urlencoded;charset=utf-8" --data-urlencode 'query=prefix sim:<http://ontology.siemens.com/si/measurement/> SELECT * WHERE { ?measurement a sim:Measurement ; sim:time_stamp "22.05.10T10:00:00.000" ; ?p ?o. } ' http://dev-sag-sms.ensure2.siemens.cloud/service/siguard-pdp-sd/getAllChannelsByTimestamp                                                                                       *   Trying 10.0.7.181...
* TCP_NODELAY set
* Connected to dev-sag-sms.ensure2.siemens.cloud (10.0.7.181) port 80 (#0)
> POST /service/siguard-pdp-sd/getAllChannelsByTimestamp HTTP/1.1
> Host: dev-sag-sms.ensure2.siemens.cloud
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Type:application/x-www-form-urlencoded;charset=utf-8
> Content-Length: 241
>
* upload completely sent off: 241 out of 241 bytes
< HTTP/1.1 400 Bad Request
< Date: Wed, 24 Aug 2022 08:51:43 GMT
< Server: Apache/2.4.52 (Debian)
< X-Powered-By: PHP/7.4.27
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
< Access-Control-Allow-Origin: *
< Content-Length: 133
< Content-Type: text/html; charset=UTF-8
< Set-Cookie: PHPSESSID=o8nm7ek0uobl9tl5ma5rkpdb96; path=/
< Connection: close
<
Erroneous request: SPARQL query with HTTP POST method but unexpected 'Content-Type': application/x-www-form-urlencoded;charset=utf-8
``````

So I have the question whether there is a way to configure the microservices in a way that they accept content-type information with charset info attached.

I added two logfiles to this issue where I added tha query command in the top.

1. sms-2022-08-24.log from a successfully operated direct service call.
2. sms-2022-08-24.service.log from a failed call using federated queries.
[sms-2022-08-24.log](https://github.com/frmichel/sparql-micro-service/files/9414608/sms-2022-08-24.log)
[sms-2022-08-24.service.log](https://github.com/frmichel/sparql-micro-service/files/9414609/sms-2022-08-24.service.log)

thanks, Christian

**Comment by kleegrewec** [comment]
Hi Grank, I am on vacation for the next 3 weeks. I will not do anything until the 19th oh September. Best regards, Christian

Von meinem iPhone gesendet

Am 26.08.2022 um 15:26 schrieb Franck Michel ***@***.***>:

﻿

Hi Christian, good one. I did not know it is possible to pass this charset within the Content-Type header.
So clearly this is a bug, easily identifiable in module frmichel/sparqlms/commonUtils.php, function getSparqlQuery().

If you want a quick and dirty fix, you can connect to the container and change the code in /sparql-micro-service/src/common/Utils.php. But of course this will be lost every time you restart the container. At least you'll know if that fixes the pb.

Here is the new code of the function. I've written it quickly but did not test it... so it may not work straight away. It replaces the internal switch with an if-then-else using str_contains.

    static public function getSparqlQuery()
    {
        $method = $_SERVER['REQUEST_METHOD'];
        switch ($method) {
            case 'GET':
                {
                    if (array_key_exists('query', $_GET))
                        $sparqlQuery = $_GET['query'];
                    else
                        self::httpBadRequest("SPARQL query with HTTP GET method but no 'query' argument.");
                    break;
                }
            case 'POST':
                {
                    if (array_key_exists('CONTENT_TYPE', $_SERVER))
                        $contentType = $_SERVER['CONTENT_TYPE'];
                    else
                        self::httpBadRequest("SPARQL query with HTTP POST method but no 'Content-Type' header.");

                    if (str_contains($contentType, 'application/x-www-form-urlencoded')) {
                        if (array_key_exists('query', $_POST))
                            $sparqlQuery = $_POST['query'];
                            else
                                self::httpBadRequest("SPARQL query with HTTP POST method and Content-Type' application/x-www-form-urlencoded' but no 'query' argument.");
                                break;
                    } else if (str_contains($contentType, 'application/sparql-query')) {
                        $sparqlQuery = file_get_contents('php://input');
                        break;
                    } else
                        self::httpBadRequest("SPARQL query with HTTP POST method but unexpected 'Content-Type': " . $contentType);
                    break;
                }
            default:
                self::httpMethodNotAllowed("Unsupported HTTP method " . $method);
        }
        return $sparqlQuery;
    }


I'll need a bit more time to test it, commit and publish a new version of the Docker image. I guess you have already produced your own version of the Docker image so you may put this change into it.

I'm pretty busy these coming days and weeks, I'll try to find a moment to do that. Meanwhile, if you try that let me know if that works.

**Comment by kleegrewec** [comment]
Hi Frank

The solution you provided solved the problem. Thanks for your help, 

Christian

**Comment by frmichel** [comment]
Ok great, thx for the hint. I've still not taken time to test it further and commit it and all. I let you know when the image is updated.

**Comment by frmichel** [comment]
Hi @kleegrewec, just to let you know, I've commited the fix and published a new release + updated Docker image v0.5.4.
